daemon/cachedump.c

```diff
@@ -851,8 +852,7 @@ int print_deleg_lookup(RES* ssl, struct worker* worker, uint8_t* nm,
 		"of %s\n", b)) 
 		return 0;
 
-	lock_rw_rdlock(&worker->env.fwds->lock);
-	dp = forwards_lookup(worker->env.fwds, nm, qinfo.qclass);
+	dp = forwards_lookup(worker->env.fwds, nm, qinfo.qclass, nolock);
```

Comment:
Both forwards_lookup and hints_lookup_stub return a pointer to the delegpt that is part of the forward and hints tree. And that has to be used while the lock is held, and so the nolock option, where it routine releases the lock and then returns a pointer to contents in the tree, is not good for thread safety. The routine has to stay as it is and protect the accesses, that print the delegation point, by holding the lock.

Reply:
But it does not release the lock when the return value is not NULL. So that solves it.